### DISCUSSION TITLE
LR2orajaとの統合について

### DISCUSSION THREAD

**Discussion by dljunji** [exch-bms2/beatoraja#668] [General]
議論の場ということですので、beatorajaとLR2orajaの統合について話してみたいです。
github初心者のため勝手が分かっていないかもしれませんがご容赦ください。
なお、後述いたしますが本議論における統合対象は判定の部分であり、LR2orajaはwcko87さんがリリースしたものを指します。

私は現在LR2orajaユーザーですが、beatorajaの主流なIRであるMocha-RepositoryとMinIRは、正式に配布されているbeatoraja以外からのスコア登録を認めていません。
したがって、Mocha-RepositoryのLevel Review機能やMinIRのStellawayとの連携機能のような独自機能は、LR2orajaでは利用ができません。
統合を考えるに至った理由として、LR2orajaでもこれらの独自機能を利用したいという要望はもちろん含まれます。
ただそれだけではなく、判定の違いからクライアントが分かれていることで、LR2からの移行を決めたユーザーが分散してしまう現状をもったいなく感じており、こと判定に限れば統合できるのではないかと考えているのです。

ここで、LR2からbeatorajaへの移行を考えているユーザーにとっての主な懸念事項は以下の2点だと考えています。

1点目は判定の違いです。
beatorajaはLR2より判定が甘いだとかクリアが簡単という話はよく見聞きします。
wcko87さんが作成したbeatorajaの派生版であるLR2orajaは、この懸念を払拭するのに一役買っており、LR2の判定に近いクライアントとして一定の支持を得ていると認識しています。
このLR2orajaには、rekidai-infoさんが機能拡張したものも存在し、一部機能はのちにbeatorajaへ取り込まれた経緯もあると認識しています。
また、最近LR2oraja専用のIRとしてtachi-beatoraja-irが作られたようです。

2点目はLR2IRが利用できないことです。
LR2を利用するにあたり、LR2IRの役割は大きく3つあると思います。
まず、長年の蓄積で多数のスコアが登録されたIRであること。
まだ歴史の浅いbeatorajaの各IRと登録スコアの絶対数を比べても仕方がないため割愛しますが、rekidai-infoさんによってbeatorajaとLR2orajaからLR2IRのデータを参照できるLR2IR Read Onlyが提供されています。
次に、LR2IRのデータを利用したWebサービスが使えること。
代表的な機能であるリコメンド機能は、本来LR2IR上に自分で登録したデータを元に算出しますが、らゆさんが作成したツールによってbeatorajaとLR2orajaどちらも擬似的に利用できます。
最後に、LR2IRに登録されたライバルの譜面を借りるG-BATTLE機能が使えること。
これは現在のbeatorajaでは利用できない機能であり、似た機能が実装されない限りは、今後も比較されLR2優位とされ続ける点だと考えています。

私がLR2orajaを利用する理由もこれらによるものですが、ユーザーが触れるクライアントはbeatorajaとして統一させた上で、各々の好みに応じた判定を選択できることが今後のコミュニティー発展のためにも理想的ではないか、というのが主張です。

以下、統合すると仮定した場合の実装案になります。
beatorajaの起動時のオプションもしくは楽曲プレー時のオプションで判定（beatoraja／LR2）を選択できるようにします。
既存IR上では、プレースタイル（キーボード／専用コントローラー）が識別できるように、判定（beatoraja／LR2）も識別できるようにします。
既存のスコアは全てbeatoraja判定であり、後からローカルDBのスコアをIRへ送信することはできないため、どちらの判定か識別できないスコア情報がIR上に混在することはないはずです。
ただし、当然この識別を行うためには、各IRや連携サービスの管理者様にもご対応いただく必要があります。
クライアント上でIRのスコアを参照する際には、IR上に登録された全ての情報を取得するかもしくは指定した判定に応じた情報のみ取得するかを、選択できるようにするのが望ましいと考えます。

皆様のご意見もお聞かせいただけますと幸いです。
長々と失礼いたしました。

**Comment by SackMagiclight** [comment]
要するに、「LR2のほぼ完全なクローンが欲しい」ってことなんですが、beatorajaはそこを目指してはいない（はず）です。
とはいえ、一部見習えるところは継承されていますが。

***
そもそも判定に関してはEXPAND JUDGEでそこそこ自由度高く拡張可能です。  
（判定幅減らす方向であればIRにも送信できますし）

**Comment by katandps** [comment]
本件は対応不要/対応すべきでないと考えます。

判定について、そもそもbeatorajaとLR2は異なるグラフィック機構の上に実装されているものであり、判定の幅を擬似的に合わせたところで厳密に同じ判定にはなりません。根本的に異なるソフトウェアです。

IRについても、上記の議論からLR2IRのデータとは比較不能なため、現存するLR2IRに無理やりつなぐような実装をするメリットはほとんどないと言えます。リプレイ再生機能を持つIRが欲しい場合はインターフェースの提案・実装を行えばよいと思います。
現行beatorajaではリプレイがあれば譜面再現ができるので、特にLR2に合わせる必要性はありません。(なお、LR2のリプレイで譜面再現はおそらく不可能です)

LR2の蓄積データをもとにした解析（実力推定など）はLR2のプレイヤーと比較することに意味はありません。また、不正/複垢の影響などにより、LR2IRのデータがどのくらい有効かは疑問視されるところです。
beatoraja側で解析可能なデータが集まるよう、判定環境を散らすべきではないのではないですか？

**Comment by dljunji** [comment]
早速のコメントありがとうございます。
拝見したところ私の付けたタイトルが不適当だったかもしれません。
2世帯住宅で玄関が2つあるように、beatorajaを起動したユーザーがbeatorajaとLR2orajaの判定を選べる形にできないか、と考えたのです。

現在のLR2orajaの利用者は、おそらくLR2と全く同じ判定でないことは重々承知の上でLR2orajaを利用しているでしょう。
私を含め彼らは判定の壁さえ取り除けば本来beatorajaの利用者となり得る層であり、コミュニティーに取り込むことが可能ではないかと思いました。

移行時の懸念事項としてLR2との機能比較はしましたが、LR2のクローン（後継機）としてのbeatorajaが欲しいだとか、beatorajaからLR2IRに繋ぎたいという要望とイコールではありません。
今後もLR2の継続利用を望むユーザーを無理やり引きずり込むような考え方とも異なります。

いかがでしょうか。
引き続き皆様のご意見お待ちしております。

**Comment by MiraiScarlet** [comment]
wckoさんがbeatorajaにLR2判定を組み込むための開発をしていたはず、と思ってましたがそれは置いておいて。

---

自分も統合に関してはやらない方がいいと思っています。

極端な話としては、(LR2(他ソフト)判定を実装したことにより)**beatoraja判定の否定に繋がりかねないのではないか**、と危惧しています。
過去もそうであったように、各BMS Player(Software)における判定解釈の有意性はあった方がいいのではないか、と考えています。

各ソフトの良し悪しをユーザーが各自判断した上で、それぞれのプレイヤーを使用すればいいのではないかと思います。

※個人的にはLR2oraja使用者をbeatorajaに取り込むことは不可能だと考えています。

**Comment by dljunji** [comment]
コメントありがとうございます。

表現を引用させていただくと、beatoraja判定が否定された結果生まれたのがLR2orajaだと思いますので、前後関係がおかしいです。
それに優位性（誤字でしょうか？）というならば一つのクライアントで好みの判定を選択できることは、選択肢の多さという点で優位にはたらくのではないでしょうか。

Mocha-Repositoryを拝見しますと、スコアに加えて楽曲情報とBMSユーザーのリポジトリを目指す、とあります。
私はこの考え方には共感していますし、beatorajaのメインストリームのIRであるからこそ、より多くのユーザーが気軽に参加できる場であって欲しいと考えています。

私の発想は、公式リリースされるbeatorajaにおいて、各ユーザーが従来のbeatoraja判定とLR2oraja判定あるいは第3、第4の判定を自由に選択できるようにすることであるため、LR2orajaの利用者を（一方的に）取り込むという解釈は誤りです。

**Comment by dljunji** [comment]
コメントありがとうございます。

表現を引用させていただくと、beatoraja判定が否定された結果生まれたのがLR2orajaだと思いますので、前後関係がおかしいです。
それに優位性（誤字でしょうか？）というならば一つのクライアントで好みの判定を選択できることは、選択肢の多さという点で優位にはたらくのではないでしょうか。

Mocha-Repositoryを拝見しますと、スコアに加えて楽曲情報とBMSユーザーのリポジトリを目指す、とあります。
私はこの考え方には共感していますし、beatorajaのメインストリームのIRであるからこそ、より多くのユーザーが気軽に参加できる場であって欲しいと考えています。

私の発想は、公式リリースされるbeatorajaにおいて、各ユーザーが従来のbeatoraja判定とLR2oraja判定あるいは第3、第4の判定を自由に選択できるようにすることであるため、LR2orajaの利用者を（一方的に）取り込むという解釈は誤りです。


※追記：読み返したところ優位性云々のくだりは私の解釈が誤っておりましたので以下追記いたします。

クライアントごとの独自性という意図かと推察します。
beatorajaや各種IRの開発者側の思想として、beatorajaとLR2orajaは完全に別物として扱い、今後一切交わることは認められないと言うのであれば、そのように明言いただければこれ以上の議論は不要かと思います。

**Comment by MiraiScarlet** [comment]
文面の修正はコメント右上のカラム(・・・の部分です)のEditから可能ですので参考までに。

---

個人的には、LR2orajaのように開発側で判定を用意するのではなく、現在のEX-JUDGEのようにユーザー側で数値をある程度自由に動かせる環境の方が良いのかなと思っています(BAD以下も動かせるようになるのがベスト？要検討)。
各判定を個別実装してしまうとデバッグ作業などに不都合が発生してしまう可能性があると思います。

ゲージの減少量や補正量を設定で厳しくしたり緩くしたりできるとなおいいのかもしれないです。
ただ、これらを実装する事でIR側の負担が増しそうな予感がしていますが…

**Comment by dljunji** [comment]
ご助言いただきありがとうございます。
アプリでは既存コメントの編集をしたつもりでしたが新規投稿になってしまいました。

EXPAND JUDGEで判定を厳しくすることは結局のところ個人の縛りプレーの範疇にとどまってしまい、LR2oraja判定として同一の判定で競い合いたい層の要求は満たせません。

デバッグ負荷の増加については、私自身がコントリビューターでは無いため多くは語りませんが、本件はクライアント側の変更に伴う今後のデバッグ量よりもIR等のメンテナンスに伴う管理側の負担が大きいと考えています。
したがいまして、先にも述べた通りそれらの管理者がbeatoraja判定とLR2oraja判定との同居を拒絶するのであれば、本議論はこれ以上の発展は無いでしょう。

**Comment by exch-bms2** [comment]
※重複発言は隠しておきました

> LR2oraja判定として同一の判定で競い合いたい

真意はおそらく、過去の自身のLR2実績、および他LR2プレイヤーと競いたい、も含まれていると思いますが、LR2でプレイしないのはなぜでしょうか？
判定のカタログ値を同一にしたところで、キー入力時の処理や判定アルゴリズムが違う以上、LR2とは違う土俵であることには変わりなく、違う土俵のスコアやランプグラフ、リコメンドを比較することは競技的にあまり意味のない行為だと考えます。
LR2の土俵で競いたいならLR2(当然ですがバイナリ改変されていないもの)でプレイするべきです。

それはともかく、

> beatorajaの起動時のオプションもしくは楽曲プレー時のオプションで判定（beatoraja／LR2）を選択できるようにします

将来的にはいくつかの他BMSプレイヤー(nazoBMplay、ナナシ、ruv-it、LR2、angolmois、Raindrop、etc...)の判定を選択してプレイできることを考えています。
IR側の対応については、別IRサーバーを立ててくれる方がいればいいんじゃないかなとは思いますが。
(正直なところ、リターンに対して負担が大きすぎる)

**Comment by dljunji** [comment]
私のミスでお手数おかけいたしました。ご配慮感謝します。

さて、皆さんやはり先入観が強いというか少し勘違いされているように思います。
単純な話です。慣れ親しんだ判定により近い判定かつ最新の環境でBMSを楽しめればそれに越したことは無いのではないでしょうか。

LR2は優れたクライアントでしたが更新が止まった以上、未来はありません。
（一方でLR2IRは貴重なデータベースとして、存続する限り価値を提供し得るでしょう。）
真の意味でLR2同士のスコアを競いたい層は、今後もbeatorajaへ移行することはないでしょう。

beatoraja上でLR2の判定を近似したアルゴリズムでプレーすることを許容する層を、公式のbeatorajaを提供する皆様が今後受け入れる可能性があるのか、確認したかったというのが正直なところです。
そして、将来的にユーザーが判定を選択できるようになる可能性が示唆されたことで、ひとまず私は納得いたしました。

LR2IRの一時停止を機に新たにbeatoraja(LR2oraja)に触れる機会を持つユーザーを目にしました。
本トピックは都合よく現れた議論の場に、純粋な疑問をぶつけてみただけなのです。
今後も開発に関わる皆様を陰ながら応援しております。

**Comment by dljunji** [comment]
先のコメントの通り私自身は意見を伝え腑落ちしましたが、反論や同調その他のご意見ございましたらお願いいたします。

**Comment by MiraiScarlet** [comment]
判定の件何処かで話してたような記憶がありましたが失念していました。
申し訳ありません。

---

同一本体でプレイするのであれば、IRでの判別機能は必須かと思われます。
そこに労力を割くかと言われたら別の機能を追加する事の方がよさそうではあります。

**Comment by DolphinDTM** [comment]
この議論スレッド全体を機械翻訳で読みました。もし私が何か誤解していたのなら、どうかお許しください。

私は、問題のいくつかの側面を解決するのに役立つ可能性がある提案を持っています。
難易度表の管理者が段位コースにLR2ゲージを強制できるので、難易度表の管理者がすべての譜面にLR2ゲージと判定を強制できるようにしたらどうでしょうか。難易度表からBMSにアクセスした場合、LR2ゲージと判定が適用されます。それ以外の場合は、デフォルトでbeatorajaのゲージと判定が適用されます。難易度表でLR2ゲージと判定が指定されていない場合、beatorajaゲージと判定がデフォルトで適用されます。

GENOCIDEをプレイしている人は、beatorajaの中のレガシー機能として使うことができます。実はこれ、GENOCIDEにしか関係ないと思うんです。GENOCIDEのレベル評価はLR2イージークリアが基準です。StellaやSatelliteはbeatorajaの判定やゲージに問題はないようです。
beatorajaさんの判断やゲージを貶めることなく、LR2さんとbeatorajaさんの橋渡しをすることができるのではないでしょうか。

ただ、懸念されるのは、実装が複雑になり、すべての責任が難易度表の管理者に委ねられてしまうのではないかということです。
あくまでアイデアなので、皆さんのご意見もお聞かせください。

**Comment by wcko87** [comment]
LR2の判定やゲージをbeatorajaに統合するのはちょっと複雑と気がします。というのも、判定やゲージ以外、プレイに影響する他の違いもあるからです。

LR2orajaとbeatorajaの相違点：
1. ゲージ
2. 判定 (EASY+NORMAL+HARD+VERYHARD, LN放す判定, 皿)
3. VERYEASY判定がない（NORMAL判定になる）
4. LR2ハードで2%以下になると失敗する（コースゲージも同じ）
5. LN放すとBAD
6. #TOTALなし譜面のTOTAL計算

結局LR2orajaがbeatoraja本体に統合されることをコミュニティが望んでいるかどうかはわかりませんし、自分も別々にすることに問題はないと思います。
現状では、統合を容易にするために、LR2orajaの判定・ゲージ差分「1+2」をリファクタリングしました ( https://github.com/wcko87/lr2oraja/commit/52b6cdf97742d86e5d909daf9d371b324c45d8ea )。しかし、「3+4+5+6」は統合するのほうがややこしいです。
そしてLR2orajaのREADME.mdで全ての差分をまとめました。https://github.com/wcko87/lr2oraja。

また、LR2orajaはユーザーがLR2IRのスコアと競争するために設計されているわけではありません。現在BMSの譜面の難易度はLR2のパラメータに合わせたので、LR2に近いものになるように心がけています。特に、beatorajaのハードゲージの回復速度とダメージの減少量、とGOODの判定、と皿の判定で、譜面の感覚が違いすぎになると思います。

**Comment by nightmare3685** [comment]
あまり詳しくはないのですが
LR2のランキングがサービス終了ということで今後はorajaにパラダイムシフトするのではないかと思います。
判定については統合はしてもしなくても良いと考えます。
なぜならば容量が許すならどちらのソフトも利用すればいいからです。
逆にどちらも使うのは煩わしいと感じる層が増えたならLR2の判定を判定を取り入れた例のソフトと協力関係になればいいと思います。
私はソフトの切り替えが面倒なので統合には賛成です
